pkg/binlog/reader/tcp.go

```diff
@@ -0,0 +1,163 @@
+
+	"github.com/pingcap/dm/pkg/gtid"
+	"github.com/pingcap/dm/pkg/log"
+	"github.com/pingcap/dm/pkg/utils"
+)
+
+// TCPReader is a binlog event reader which read binlog events from a TCP stream.
+type TCPReader struct {
+	stageMu sync.Mutex
+	stage   readerStage
+
+	syncerCfg replication.BinlogSyncerConfig
+	syncer    *replication.BinlogSyncer
+	streamer  *replication.BinlogStreamer
+}
+
+// TCPReaderStatus represents the status of a TCPReader.
+type TCPReaderStatus struct {
+	Stage      string `json:"stage"`
+	Connection uint32 `json:"connection"`
+}
+
+// String implements Stringer.String.
+func (s *TCPReaderStatus) String() string {
+	data, err := json.Marshal(s)
+	if err != nil {
+		log.Errorf("[TCPReaderStatus] marshal status to json error %v", err)
+	}
+	return string(data)
+}
+
+// NewTCPReader creates a TCPReader instance.
+func NewTCPReader(syncerCfg replication.BinlogSyncerConfig) Reader {
+	return &TCPReader{
+		syncerCfg: syncerCfg,
+		syncer:    replication.NewBinlogSyncer(syncerCfg),
+	}
+}
+
+// StartSyncByPos implements Reader.StartSyncByPos.
+func (r *TCPReader) StartSyncByPos(pos gmysql.Position) error {
+	r.stageMu.Lock()
+	defer r.stageMu.Unlock()
+
+	if r.stage != stageNew {
+		return errors.NotValidf("stage %d, expect %d", r.stage, stageNew)
+	}
+
+	streamer, err := r.syncer.StartSync(pos)
+	if err != nil {
+		return errors.Annotatef(err, "start sync from position %s", pos)
+	}
+
+	r.streamer = streamer
+	r.stage = stagePrepared
+	return nil
+}
+
+// StartSyncByGTID implements Reader.StartSyncByGTID.
+func (r *TCPReader) StartSyncByGTID(gSet gtid.Set) error {
+	r.stageMu.Lock()
+	defer r.stageMu.Unlock()
+
+	if r.stage != stageNew {
+		return errors.NotValidf("stage %d, expect %d", r.stage, stageNew)
+	}
+
+	if gSet == nil {
+		return errors.NotValidf("nil GTID set")
+	}
+
+	streamer, err := r.syncer.StartSyncGTID(gSet.Origin())
+	if err != nil {
+		return errors.Annotatef(err, "start sync from GTID set %s", gSet)
+	}
+
+	r.streamer = streamer
+	r.stage = stagePrepared
+	return nil
+}
+
+// Close implements Reader.Close.
+func (r *TCPReader) Close() error {
+	r.stageMu.Lock()
+	defer r.stageMu.Unlock()
+
+	if r.stage == stageClosed {
+		return errors.NotValidf("already closed")
+	}
+
+	connID := r.syncer.LastConnectionID()
+	if connID > 0 {
+		dsn := fmt.Sprintf("%s:%s@tcp(%s:%d)/?charset=utf8mb4",
+			r.syncerCfg.User, r.syncerCfg.Password, r.syncerCfg.Host, r.syncerCfg.Port)
+		db, err := sql.Open("mysql", dsn)
+		if err != nil {
+			return errors.Annotate(err, "open connection to the master")
+		}
+		defer db.Close()
+		err = utils.KillConn(db, connID)
+		if err != nil {
+			return errors.Annotatef(err, "kill connection %d", connID)
+		}
+	}
+
+	r.stage = stageClosed
+	return nil
+}
+
+// GetEvent implements Reader.GetEvent.
+func (r *TCPReader) GetEvent(ctx context.Context, checkStage bool) (*replication.BinlogEvent, error) {
+	if checkStage {
+		r.stageMu.Lock()
+		if r.stage != stagePrepared {
```

Comment:
would atomic operation be better?

Reply:
changed to use atomic operation, and removed `checkStage`.